Implement special config file to dump parameters as json.

jmchilton opened the pull request:
```
    <configfiles>
        <inputs name="inputs" />
    </configfiles>
```

Version is specified as 1, because the Galaxy tool plumbing doesn't really distinguish between parameter is '', parameter was explicitly supplied as None, parameter was absent, etc... - everything is a just a string in some ways. This distinction may be more important with JSON and for instance the CWL will require more precision, so we may want to revise Galaxy's tool handling in a backward compatible way someday.

Both @erasche  and @bgruening are working on tools that seem they would benefit from this.

jmchilton committed:
Implement special config file to dump parameters as json.

```
    <configfiles>
        <inputs name="inputs" format="json" version="1" />
    </configfiles>
```

Version is specified as 1, because the Galaxy tool plumbing doesn't really distinguish between parameter is '', parameter was explicitly supplied as None, parameter was absent, etc... - everything is a just a string in some ways. This distinction may be more important with JSON and for instance the CWL will require more percision, so we may want to revise Galaxy's tool handling in a backward compatible way someday.

hexylena commented:
Will look at this tonight if @bgruening doesn't get to it first.

jmchilton committed:
More work on converting tool inputs to json.

 - Drop version requirement per commends by @bgruening.
 - Add examples for conditionals, repeats, sections, colors, and booleans specified truevalue/falsevalue.
 - Various bug fixes unearthed by new test cases.

jmchilton committed:
Update inputs_as_json.xml to drop version attribute.

Also add note about filename.

bgruening commented on test/functional/tools/inputs_as_json.xml:
```diff
@@ -0,0 +1,79 @@
+    assert_equals(as_dict["test_case"], 2)
+    assert_equals(as_dict["text_test"], "bar")
+    assert_equals(as_dict["booltest"], False)
+    assert_equals(as_dict["inttest"], None)
```
Not sure if this should be inlcuded in the dict at all. Both have advantages and disadvantages but I would assume if a parameter is not set, it does not appear in my config and I don't need to deal with it.

jmchilton replied:
I can't distinguish between an API call that left this unset and it appearing on a form and the user specifying `''`. So I would prefer to leave it this way.

bgruening commented on test/functional/tools/inputs_as_json.xml:
```diff
@@ -0,0 +1,79 @@
+<tool id="inputs_as_json" name="inputs_as_json" version="1.0.0">
+    <configfiles>
+        <inputs name="inputs" format="json" version="1" />
+        <configfile name="check_inputs"><![CDATA[
```
This one will fail planemo lint, because of the ordering. :)

jmchilton replied:
All these tools will fail... but I will fix the ordering.

bgruening commented on test/functional/tools/inputs_as_json.xml:
```diff
@@ -0,0 +1,79 @@
+        <param name="test_case" type="integer" value="0" />
+        <param name="text_test" type="text" />
+        <param name="booltest" truevalue="booltrue" falsevalue="boolfalse" checked="false" type="boolean" />
+        <param name="inttest" type="integer" optional="true" />
```
For documentation purpose can we include a conditional here?
Especially in the case of a conditional the we don't want to include a large list of `None`s, isn't it?

bgruening commented:
John I don't like the version attribute in a single param/feature. I guess we should collect those ideas and increase/version the entire Galaxy Tool config at some point. Or do we really want to end up with different params with different version numbers?

Great work, thanks! I guess we own you now something, I haven't followed yesterdays IRC closely ;)

jmchilton committed:
Expose hidden parameters in inputs json.

jmchilton commented:
Updated with a second commit de0c913.
- Dropped version requirement per commends by @bgruening.
- Added examples for conditionals, repeats, sections, colors, and booleans specified using truevalue/falsevalue instead of true/false.
- Various bug fixes unearthed by new test cases.

hexylena commented:
As soon as I add the `<inputs name="inputs" />` to my tool XML, I see the following on trying to run the tool:

![selection_522](https://cloud.githubusercontent.com/assets/458683/12056463/46c0813a-aefc-11e5-85a4-127c565d2843.png)

Edit: whoops, sorry, hit "Close PR" instead of comment. My apologies.

jmchilton commented:
@erasche Link?

hexylena commented:
Link to what?

hexylena commented:
Beautiful, this would've saved me probably a dozen hours debugging configfiles, over the course of every version of my jbrowse tool.

![selection_523](https://cloud.githubusercontent.com/assets/458683/12056598/c5b40808-aefd-11e5-885e-2eb748576266.png)

Huge :+1: from me.

bgruening commented:
Awesome!
